--- hunspell-1.2.2.orig/src/hunspell/hunspell.hxx	2008-04-02 13:12:03.000000000 +0100
+++ hunspell-1.2.2/src/hunspell/hunspell.hxx	2008-05-16 14:49:12.000000000 +0100
@@ -51,7 +51,8 @@
    * input: path of affix file and dictionary file
    */
 
-  Hunspell(const char * affpath, const char * dpath, const char * key = NULL);
+  Hunspell(const char * affpath, const char * dpath);
+  Hunspell(const char * affpath, const char * dpath, const char * key);
   ~Hunspell();
 
   /* load extra dictionaries (only dic files) */
@@ -170,7 +171,7 @@
    int    insert_sug(char ***slst, char * word, int ns);
    void   cat_result(char * result, char * st);
    char * stem_description(const char * desc);
-
+   void init(const char * affpath, const char * dpath, const char * key);
 };
 
 #endif
--- hunspell-1.2.2.orig/src/hunspell/hunspell.cxx	2008-04-02 13:12:30.000000000 +0100
+++ hunspell-1.2.2/src/hunspell/hunspell.cxx	2008-05-16 14:49:50.000000000 +0100
@@ -22,8 +22,18 @@
 #endif
 #endif
 
+Hunspell::Hunspell(const char * affpath, const char * dpath)
+{
+    init(affpath, dpath, NULL);
+}
+
 Hunspell::Hunspell(const char * affpath, const char * dpath, const char * key)
 {
+    init(affpath, dpath, key);
+}
+
+void Hunspell::init(const char * affpath, const char * dpath, const char * key)
+{
     encoding = NULL;
     csconv = NULL;
     utf8 = 0;
